### PR TITLE
FIx Finding Housekeeping Candidates and Modify Housekeeping Structure

### PR DESCRIPTION
**Pull request by devleejb**
**What this PR does / why we need it**:

- Refactored the `FindDeactivateCandidates` function, moving it from the Database Layer to the Service Layer
    - Due to the change in the function's layer, the Housekeeping-related tests from the Database Layer have been relocated to Integration Tests
- Changed function name from `listProjectInfos`  to `FindNextNCyclingProjectInfos`
- Changed the implementation of the `FindNextNCyclingProjectInfos` function to retrieve a cyclic list of N projects with IDs greater than lastProjectID
- Fix `FindDeactivateCandidatesPerProject` for memdb to return candidates of only one project

**Which issue(s) this PR fixes**:

Fixes # https://github.com/yorkie-team/yorkie/issues/744

**Special notes for your reviewer**:

- Considering whether initializing the database in the Integration Test is the right approach.
- Concerns arise due to the absence of clear code for each Unit Test, potentially leading to data dependency issues.
- There is a remaining function `FindDeactivateCandidatesPerProject` in the Database Layer. Seeking opinions on whether it would be appropriate to rename it to something unrelated to Housekeeping.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation**:

```docs

```

**Checklist**:
- [x] Added relevant tests or not required
- [x] Didn't break anything

### CODE REVIEW

File: server/backend/database/mongo/client.go [hunk 1]
```diff
@@ -268,6 +268,28 @@ func (c *Client) listProjectInfos(
 		return nil, fmt.Errorf("fetch project infos: %w", err)
 	}
 
+	if len(infos) < pageSize {
+		opts.SetLimit(int64(pageSize - len(infos)))
+		if err != nil {
```
Comment (krapie) [inline server/backend/database/mongo/client.go:273]: Is there any reason for this error check?

File: server/backend/database/memory/database.go [hunk 2]
```diff
@@ -247,7 +247,19 @@ func (d *DB) listProjectInfos(
 	for i := 0; i < pageSize; i++ {
 		raw := iter.Next()
 		if raw == nil {
-			break
+			iter, err = txn.LowerBound(
```
Comment (krapie) [inline server/backend/database/memory/database.go:250]: Does this query repeats for `pageSize` - `fetched projects size`?
We might need to improve these codes.

File: server/backend/database/memory/database.go [hunk 3]
```diff
@@ -224,8 +224,8 @@ func (d *DB) CreateProjectInfo(
 	return info, nil
 }
 
-// listProjectInfos returns all project infos rotationally.
-func (d *DB) listProjectInfos(
+// rotateListProjectInfosGreaterThan returns all project infos rotationally.
+func (d *DB) rotateListProjectInfosGreaterThan(
```
Comment (krapie) [inline server/backend/database/memory/database.go:228]: How about naming this function to just `listProjectInfosGreaterThan()` or other names? For me, the word `rotate` seems ambiguous.

**Review by krapie — COMMENTED** [review]
> I'm curious if there are ways to leverage the features of the Go language to further improve the code in a cleaner way.

I can't think of anything beyond what you've written, other than using a for statement to 
simply code (maybe?) or refactoring redundant querying codes. I didn't find any syntax for circular queries in MongoDB either.

We can temporary apply this changes to fix urgent error, and then consider housekeeping service structure to improve overall housekeeping structure & performance.

Any ideas? @hackerwins

File: server/backend/database/mongo/client.go (hunk 1 above)
Comment (devleejb) [reply server/backend/database/mongo/client.go]: It's useless error checking. I removed it.

File: server/backend/database/memory/database.go (hunk 2 above)
Comment (devleejb) [reply server/backend/database/memory/database.go:255]: I have made the adjustments according to your feedback. I have refactored the code to use common logic without re-invoking `raw = iter.Next()` within the if statement. However, the impact of the improvement seems to be limited. If you have any suggestions or opinions, please share it.

File: server/backend/database/memory/database.go (hunk 3 above)
Comment (devleejb) [reply server/backend/database/memory/database.go]: Sorry, It is my mistake.

I changed the function name `rotateListProjectInfosGreaterThan` to `nextCyclingProjectInfos`. How about this?

**Comment by devleejb** [comment]
@krapie 

> I can't think of anything beyond what you've written, other than using a for statement to simply code (maybe?) or refactoring redundant querying codes. I didn't find any syntax for circular queries in MongoDB either.


I'm unfamiliar with Go lang, and I had concerns about whether the code could be simplified using techniques specific to the language. Thank you.

File: server/backend/database/memory/database.go (hunk 3 above)
Comment (krapie) [reply server/backend/database/memory/database.go]: Looks good to me 😄

**Review by krapie — COMMENTED** [review]
Can you request review after you open this PR? It seems like this is still in progress.

File: server/backend/database/testcases/testcases.go [hunk 4]
```diff
@@ -1074,3 +1026,86 @@ func RunIsDocumentAttachedTest(t *testing.T, db database.Database, projectID typ
...
+
+			assert.Equal(t, projects[((i+1)*pageSize-1)%projectCnt].ID, lastProjectID)
+		}
+
+	})
+}
+
+// RunFindDeactivateCandidatesPerProjectTest runs the FindNextNCyclingProjectInfos tests for the given db.
+func RunFindDeactivateCandidatesPerProjectTest(t *testing.T, db database.Database) {
+	t.Run("`FindDeactivateCandidatesPerProject` should search candidates correctly", func(t *testing.T) {
```
Comment (krapie) [inline server/backend/database/testcases/testcases.go:1065]: > Consider matching test name convention with other test names for now.
ex) "FindDeactivateCandidatesPerProject candidate search test"

Same here.

File: server/backend/database/testcases/testcases.go [hunk 5]
```diff
@@ -1074,3 +1026,86 @@ func RunIsDocumentAttachedTest(t *testing.T, db database.Database, projectID typ
 		assert.False(t, attached)
 	})
 }
+
+// RunFindNextNCyclingProjectInfosTest runs the FindNextNCyclingProjectInfos tests for the given db.
+func RunFindNextNCyclingProjectInfosTest(t *testing.T, db database.Database) {
+	t.Run("`RunFindNextNCyclingProjectInfos` should search projects cyclic", func(t *testing.T) {
```
Comment (krapie) [inline server/backend/database/testcases/testcases.go:1032]: Consider matching test name convention with other test names for now.
ex) "FindDeactivateCandidatesPerProject candidate search test"

File: test/integration/housekeeping_test.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	db := setupTest(t)
+
+	projects := createProjects(t, db)
+
+	coordinator := memory.NewCoordinator(nil)
+
+	h, err := housekeeping.New(config.Housekeeping, db, coordinator)
+	assert.NoError(t, err)
+
+	t.Run("`FindDeactivateCandidates` should return correct last projectID", func(t *testing.T) {
```
Comment (krapie) [inline test/integration/housekeeping_test.go:74]: > Consider matching test name convention with other test names for now.
ex) "FindDeactivateCandidatesPerProject candidate search test"

Same here.

File: test/integration/housekeeping_test.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+			ctx,
+			0,
+			fetchSize,
+			lastProjectID,
+		)
+		assert.NoError(t, err)
+		assert.Equal(t, projects[fetchSize-(len(projects)%3)-1].ID, lastProjectID)
+	})
+
+	t.Run("`FindDeactivateCandidates` should return correct clients", func(t *testing.T) {
```
Comment (krapie) [inline test/integration/housekeeping_test.go:101]: > Consider matching test name convention with other test names for now.
ex) "FindDeactivateCandidatesPerProject candidate search test"

Same here.

File: test/integration/housekeeping_test.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+			ctx,
+			10,
+			10,
+			database.DefaultProjectID,
+		)
+
+		assert.NoError(t, err)
+		assert.Len(t, candidates, 2)
+		assert.Equal(t, candidates[0].ID, clientA.ID)
+		assert.Contains(t, candidates[1].ID, clientB.ID)
```
Comment (krapie) [inline test/integration/housekeeping_test.go:131]: How about using only `assert.Equal(t, candidates[1].ID, clientB.ID)` or `assert.Contains(t, candidates, clientA)` to keep consistency?

File: test/integration/housekeeping_test.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+		assert.NotContains(t, candidates, clientC)
+	})
+}
+
+func createProjects(t *testing.T, db *mongo.Client) []*database.ProjectInfo {
+	t.Helper()
```
Comment (krapie) [inline test/integration/housekeeping_test.go:137]: Using `t.Helper()` function looks good to me 😄

**Review by krapie — CHANGES_REQUESTED** [review]
Thank you for your contribution.

Refactored housekeeping service structure & test looks better than before 👍🏼 
I have requested few convention-related changes.

**Review by krapie — APPROVED** [review]
LGTM.

**Comment by krapie** [comment]
@devleejb 
Leaving few more comments on your notes:

> Considering whether initializing the database in the Integration Test is the right approach.

initializing on top of the test function looks better than `rentention_test.go`.

> Concerns arise due to the absence of clear code for each Unit Test, potentially leading to data dependency issues.

`RunFindNextNCyclingProjectInfosTest()` and `RunFindDeactivateCandidatesPerProjectTest()` test codes in `testcases.go` will do its job for now.

> There is a remaining function FindDeactivateCandidatesPerProject in the Database Layer. Seeking opinions on whether it would be appropriate to rename it to something unrelated to Housekeeping.

This function can be only used for housekeeping purposes, and cannot normalize more. So I think name `FindDeactivateCandidatesPerProject()` is appropriate.

**Pull request merged**